### PR TITLE
Issue #7446: ensure all checks summaries in sync with checks.xml

### PR DESCRIPTION
**Pull request by pbludov**
Issue #7446

`Node.getAttributes().getNamedItem("name").getNodeValue()` extracted to an utility method.

Method `sanitizeXml` moved to the class `XmlUtil`.

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 1]
```diff
@@ -258,6 +261,65 @@ private static boolean isPresent(String availableChecks, String checkName) {
         return availableChecks.matches(linkPattern);
     }
 
+    private static Map<String, String> readSummaries() throws Exception {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: Utility methods should go after the method(s) that use them.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 2]
```diff
@@ -258,6 +261,65 @@ private static boolean isPresent(String availableChecks, String checkName) {
...
+    @Test
+    public void testAllChecksPageInSyncWithChecksSummaries() throws Exception {
+        final Pattern endOfSentence = Pattern.compile("(.*?\\.)\\s", Pattern.DOTALL);
+        final Map<String, String> summaries = readSummaries();
+
+        for (Path path : XdocUtil.getXdocsConfigFilePaths(XdocUtil.getXdocsFilePaths())) {
+            final String fileName = path.getFileName().toString();
+            if ("config_reporting.xml".equals(fileName)
+                    || "config_filefilters.xml".equals(fileName)
+                    || "config_filters.xml".equals(fileName)) {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: @romani Is there any reason filters don't have their own summary page?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 3]
```diff
@@ -258,6 +261,65 @@ private static boolean isPresent(String availableChecks, String checkName) {
...
+                final Node section = sources.item(position);
+                final String sectionName = XmlUtil.getNodeName(section);
+                if (!"Description".equals(sectionName)) {
+                    continue;
+                }
+
+                final String checkName = XmlUtil.getNodeName(section.getParentNode());
+                final Matcher matcher = endOfSentence.matcher(section.getTextContent());
+                assertWithMessage(
+                    "XDocs description should start with a sentence that ends with a period")
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: Assertions require the page that is being examined, and the check that is being looked at. This is extremely helpful in debugging a failure.
See https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1587 as an example. All tests in this class should have assertions like this.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 4]
```diff
@@ -258,6 +261,65 @@ private static boolean isPresent(String availableChecks, String checkName) {
...
+                    .that(matcher.find());
+                final String firstSentence = XmlUtil.sanitizeXml(matcher.group(1));
+                assertWithMessage("Summary for check " + checkName
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: Please sync the assertion message. Also `checkName` is in the message twice.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XmlUtil.java [hunk 5]
```diff
@@ -132,4 +132,30 @@ else if (child.hasChildNodes()) {
         return result;
     }
 
+    /**
+     * Returns the value of the "name" attribute for the given node.
+     *
+     * @param node to retrieve the name
+     * @return the node name
+     */
+    public static String getNodeName(Node node) {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XmlUtil.java]: > getNodeName
> Returns the value of the "name" attribute 

Shouldn't this be something like `getNodeNameAttributeText`? Maybe something a little bit shorter, but I don't think the current name expressly says enough about what it is doing.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XmlUtil.java [hunk 6]
```diff
@@ -132,4 +132,30 @@ else if (child.hasChildNodes()) {
...
+        return node.getAttributes().getNamedItem("name").getNodeValue();
+    }
+
+    /**
+     * Removes all whitespaces at the beginning and at the end of the string and
+     * replaces repeated whitespaces in the middle of the string with a single space.
+     * Replaces XML entities with escaped values.
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XmlUtil.java]: The first sentence should be the summary, but you have it split into 2 and is a bit long.

I think the first sentence should be something like "This santizes the given string for safe use in XML documents." and then list what it specifically does as consecutive sentences.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 1 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 3 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 4 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XmlUtil.java (hunk 5 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XmlUtil.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XmlUtil.java (hunk 6 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/XmlUtil.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 2 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: No good reason, but they already all at one page https://checkstyle.org/config_filters.html and one more page for file filters

**Pull request merged**